controllers/pvc_mutator_test.go

```diff
@@ -0,0 +1,207 @@
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	pvcmutator "github.com/storageos/api-manager/controllers/pvc-mutator"
+)
+
+// SetupPVCMutatorTest will set up a testing environment.  It must be called
+// from each test.
+func SetupPVCMutatorTest(ctx context.Context, mutators []pvcmutator.Mutator) {
+	var cancel func()
+
+	BeforeEach(func() {
+		ctx, cancel = context.WithCancel(ctx)
+
+		// Configure the certificate manager.
+		certOpts := cert.Options{
+			Service: &admissionv1.ServiceReference{
+				Name:      webhookServiceName,
+				Namespace: webhookServiceNamespace,
+			},
+			Client:                    k8sClient,
+			SecretRef:                 &types.NamespacedName{Name: webhookSecretName, Namespace: webhookSecretNamespace},
+			MutatingWebhookConfigRefs: []types.NamespacedName{{Name: webhookMutatingConfigName}},
+		}
+
+		err := cert.NewManager(nil, certOpts)
+		Expect(err).NotTo(HaveOccurred(), "unable to provision certificate")
+
+		webhookInstallOptions := &testEnv.WebhookInstallOptions
+		mgr, err := ctrl.NewManager(cfg, ctrl.Options{
+			Host:               webhookInstallOptions.LocalServingHost,
+			Port:               webhookInstallOptions.LocalServingPort,
+			CertDir:            webhookInstallOptions.LocalServingCertDir,
+			LeaderElection:     false,
+			MetricsBindAddress: "0",
+		})
+		Expect(err).NotTo(HaveOccurred(), "failed to create manager")
+
+		decoder, err := admission.NewDecoder(mgr.GetScheme())
+		Expect(err).NotTo(HaveOccurred(), "failed to create decoder")
+
+		pvcMutator := pvcmutator.NewController(mgr.GetClient(), decoder, mutators)
+		mgr.GetWebhookServer().Register(webhookMutatePVCsPath, &webhook.Admission{Handler: pvcMutator})
+		Expect(err).NotTo(HaveOccurred(), "failed to setup controller")
+
+		go func() {
+			err := mgr.Start(ctx)
+			Expect(err).NotTo(HaveOccurred(), "failed to start manager")
+		}()
+
+		// Wait for manager to be ready.
+		time.Sleep(managerWaitDuration)
+	})
+
+	AfterEach(func() {
+		cancel()
+	})
+}
+
+var _ = Describe("PVC Mutator controller", func() {
+	// Define utility constants for object names and testing timeouts/durations
+	// and intervals.
+	const (
+		timeout  = time.Second * 1
+		duration = time.Second * 1
+		interval = time.Millisecond * 250
+	)
+
+	genPVC := func() corev1.PersistentVolumeClaim {
+		volumeMode := v1.PersistentVolumeFilesystem
+		return corev1.PersistentVolumeClaim{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "pvc-" + randStringRunes(5),
+				Namespace: "default",
+			},
+			Spec: corev1.PersistentVolumeClaimSpec{
+				AccessModes: []corev1.PersistentVolumeAccessMode{v1.PersistentVolumeAccessMode("ReadWriteOnce")},
+				Resources: corev1.ResourceRequirements{
+					Requests: corev1.ResourceList{
+						corev1.ResourceStorage: resource.MustParse("1Gi"),
+					},
+				},
+				VolumeMode: &volumeMode,
+			},
+		}
+	}
+
+	ctx := context.Background()
+
+	Context("When the PVC Mutator has no mutators", func() {
```

Comment:
Looks similar to the tests that we have for pod mutators. The only difference is in the mutator interface's mutate function signature. I guess we can avoid all these with a common package that provides such mutation/validation chaining by accepting something generic like a client.Object and the core logic is tested just once.

Reply:
Yes, that would be better.  Once we add a common package we can refactor, but for now I think testing this explicitly is good.